source/fundamentals/crud/write-operations/upsert.txt

```diff
@@ -4,8 +4,105 @@ Insert or Update in a Single Operation
 
 .. default-domain:: mongodb
 
-If your application stores and modifies data in MongoDB, you probably use
-insert and update operations. In certain workflows, you may need to choose
-between an insert and update depending on whether the document exists.
-In these cases, you can streamline your application logic by using the
-``upsert`` option available in the following methods:
+Overview
+--------
+
+You often use insert and update operations for applications that store
+and modify data. Sometimes, you need to choose between an insert or
+update depending on whether the document exists. 
+
+To insert or update documents, pass a query filter to the 
+:java-docs:`updateOne() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateOne(org.bson.conversions.Bson,org.bson.conversions.Bson,com.mongodb.client.model.UpdateOptions)>`, 
+:java-docs:`updateMany() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateMany(org.bson.conversions.Bson,org.bson.conversions.Bson,com.mongodb.client.model.UpdateOptions)>`
+or :java-docs:`replaceOne() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#replaceOne(org.bson.conversions.Bson,TDocument,com.mongodb.client.model.ReplaceOptions)>`
+methods. 
+
+These methods update documents that match your query filter,
+but inserts a document if the there are no matches when you set the
+upsert to true. When multiple documents match the query filter supplied to
+``updateOne()`` or ``replaceOne()``, the first document that matches
+gets updated.
+
+Specify an upsert:
+~~~~~~~~~~~~~~~~~~
+
+Use :java-docs:`UpdateOptions </apidocs/mongodb-driver-core/com/mongodb/client/model/UpdateOptions.html>`
+as the third parameter to the ``updateOne()`` and ``updateMany()``
+methods.
+
+Use :java-docs:`ReplaceOptions </apidocs/mongodb-driver-core/com/mongodb/client/model/ReplaceOptions.html>`
+as the third parameter to the ``replaceOne()`` method.
+
+Example
+-------
+
+The following example is about a paint store that sells eight different
+colors of paint. The store had their annual online sale resulting in the
+following documents in their ``paint_inventory`` collection: 
```

Comment:
**question:** Are these totals before or after the sale? It would help to clarify that.

**suggestion:** Refactor to separate the "human" language from the "programming" language. That is, state the situation with the store. Then say something like, `The store's database now shows the following...`